### ISSUE TITLE
prefer-nullish-coalescing autofix smashes boolean LHS to `??` in full-file context (type-aware)

### ISSUE THREAD

**Issue by mjn298** [oxc-project/tsgolint#950]
## Summary

`oxlint --type-aware --fix --fix-suggestions` converts `||` to `??` when both operands are non-nullable booleans (e.g. `Array.prototype.some()` returns). This silently changes semantics: with `??`, the expression returns the left operand even when it is `false`, breaking OR-chain logic.

typescript-eslint's `prefer-nullish-coalescing` correctly skips pure-boolean LHS, since the rule's purpose is to flag `T | null | undefined` LHS where `??` is safer than `||`. Pure boolean LHS does not match that target. See [typescript-eslint docs](https://typescript-eslint.io/rules/prefer-nullish-coalescing/) for the option semantics.

## Versions

- `oxlint`: 1.63.0
- `oxlint-tsgolint`: 0.22.1

## Reproducer

The bug **does not trigger** in isolation. Running `oxlint --type-aware --fix --fix-suggestions` on the following standalone file leaves it unchanged (correct):

```ts
type Item = { tag: string };
type Container = {
  outer: Array<{ items: Item[] } | undefined>;
  items: Item[];
  otherSet: Array<{ key: string }>;
};

export const probe = (c: Container, target: string): void => {
  const innerItems = c.outer[0]?.items ?? [];
  const topLevelItems = c.items;
  const hasFlagA =
    innerItems.some((x) => x.tag === "A") ||
    topLevelItems.some((x) => x.tag === "A");
  const hasFlagB = c.otherSet.some((x) => x.key === target);
  if (!(hasFlagB || hasFlagA)) throw new Error("invariant");
};
```

The bug **does trigger** when the same pattern lives inside a larger spec file (~900 lines, real imports from service modules whose return types are `z.infer<...>` of Zod schemas, the assertion lives inside a test-runner callback like `test("name", async ({ expect }) => { ... })`). On a file with that shape, `--fix --fix-suggestions` rewrites both `||` to `??`.

Wrapping the standalone snippet in a test-runner callback alone does not reproduce.

## Hypothesis

This suggests tsgolint loses type narrowing for `.some()` return values under certain import/inference complexity, treating the LHS as nullable in the autofix code path even though `--type-aware` is set. The lint check itself may be reading a different (correct) type than the autofix is, or the autofix path may not consult the type checker at all.

## Expected behavior

Match typescript-eslint: do not fire (and especially do not autofix) when the LHS is statically `boolean`. The rule should only target `T | null | undefined` LHS.

## Impact

This is a semantics-changing autofix that looks safe. The output passes lint but changes runtime behavior, e.g. `expect(a || b).toBe(true)` becomes `expect(a ?? b).toBe(true)` — the assertion now passes when `a === false` regardless of `b`. Easy to miss in code review.

## Workarounds

- Inline `// oxlint-disable-next-line typescript/prefer-nullish-coalescing` at each site.
- Exclude `prefer-nullish-coalescing` from the rules autofixed by lint-staged / pre-commit (run it as a manual-only fix).
- Skip `--fix-suggestions` and only apply safe `--fix`.

Happy to help bisect the file-context factor if useful — let me know what kind of intermediate repro would help.

**Comment by camc314** [comment]
> The bug does not trigger in isolation. Running oxlint --type-aware --fix --fix-suggestions on the following standalone file leaves it unchanged (correct):

this is going to be really hard to debug or fix this without a reproducion - DM me on X or Discord if you don't want to share the context publically on GitHub.


> Match typescript-eslint: do not fire (and especially do not autofix) when the LHS is statically boolean. The rule should only target T | null | undefined LHS.

I just tried this locally and I get the following:

<img width="410" height="209" alt="Image" src="https://github.com/user-attachments/assets/c125dab1-5147-4bb1-8e93-232a108bcdab" />

<img width="1078" height="209" alt="Image" src="https://github.com/user-attachments/assets/523d36a6-60d1-4f00-99fb-cd7bfd233dd4" />

which behaviour aligns with eslint.

If you hover over inside the editor what is the type of the LHS?

If you run with `--type-check` do you get any errors

**Comment by Sysix** [comment]
> `oxlint --type-aware --fix --fix-suggestions` converts `||` to `??` when both operands are non-nullable booleans (e.g. `Array.prototype.some()` returns). This silently changes semantics: with `??`, the expression returns the left operand even when it is `false`, breaking OR-chain logic.

You are fixing suggestions too, which can be sometimes dangerous. If you can create a reproduction-repo we could maybe fix it or mark it as dangerous too 🤞 (note: tsgolint does not have the concept of "dangerous", but oxlint does)

> typescript-eslint's `prefer-nullish-coalescing` correctly skips pure-boolean LHS, since the rule's purpose is to flag `T | null | undefined` LHS where `??` is safer than `||`. 

The ESLint rule only runs `--fix` equivalent. Suggestions are only be shown for editors quick fix menu.

**Comment by mjn298** [comment]
cool - I should probably not run `--fix-suggestions` in such a cavalier manner.

i'm sure it's obvious but my original issue is LLM slop, let me dig around and clarify further.

**Comment by Sysix** [comment]
@mjn298 looks like it will still be fixed with `--fix`. @camc314 created a PR in #951 to mark them as suggestions, aligning how `typescript-eslint` reports them.

**Comment by camc314** [comment]
Hey, let us know if there's anything missing here, but i'm going to close this as not planned.

We've altered the behaviour of this rule to be a suggestion instead of fix to indicate that it _may_ break your code.

Let me know if there's anything else - thanks!

**Comment by mjn298** [comment]
thanks - "We've altered the behaviour of this rule to be a suggestion instead of fix to indicate that it may break your code." is an ideal solution.
